proposals/0066-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,245 @@
+features requested for activation.
+- Feature-Queue account: The PDA under the Feature Gate program used to track
+features queued for activation that must have support signaled by nodes.
+
+## Detailed Design
+
+Before this SIMD is published, we will deploy a BPF program to manage feature
+activations, which will allow us to revoke pending feature activations
+(previously not possible).
+The feature activation process will otherwise be unchanged, so no SIMD is
+required.
+This program will be upgradeable; the authority and process through which it
+will be upgraded is outside the scope of this SIMD. Since this SIMD requires an
+update to the program, the upgrade details will be determined in parallel.
+
+The initial Feature Gate Program can be found in this pull request to the
+Solana Program Library:
+<https://github.com/solana-labs/solana-program-library/pull/4987>.
+The runtime changes and feature activation that will enable this program can be
+found in this feature tracking issue:
+<https://github.com/solana-labs/solana/issues/33547>.
+
+### Overview
+
+Consider two arbitrary versions of the Solana runtime, where the newer version
+has more features in its feature set.
+
+| v1.14.25  | v1.16.0   |
+| --------- | --------- |
+| Feature A | Feature A |
+|           | Feature B |
+| Feature C | Feature C |
+|           | Feature D |
+
+The runtime should activate each of these features when a strong majority of
+the network supports it. “Support” for a feature means that a node is running a
+version of the Solana runtime where the code required for the feature is
+present. For instance, on a cluster with 50% of stake running v1.14.25 and 50%
+of stake running v1.16.0, only Features A and C should be eligible for
+activation.
+
+This SIMD proposes 3 steps over the span of one epoch to determine which
+feature gates are queued for activation and assess stake support:
+
+1. On an epoch boundary, the runtime generates a list of feature-gates queued
+for activation.
+2. Validators signal which of the queued feature-gates they support.
+3. On the next epoch boundary, the runtime activates the feature-gates that
+have the necessary stake support
+
+### Representing the Current Feature Set
+
+As mentioned above, a BPF program is being deployed that will replace the
+non-existent account `Feature111111111111111111111111111111111111`, which is
+the owner of all feature accounts.
+
+The Feature Gate program will utilize two PDAs to track the authoritative
+feature set queued for activation. One PDA (the Feature-Request PDA) will
+store newly requested feature activations and the other (the Feature-Queue
+PDA) will store the previous epoch’s requested activations.
+
+When a new epoch begins, the Feature-Request PDA will be an empty list. When
+key-holders submit the CLI command to activate a feature, the Feature Gate
+Program appends the feature ID to the Feature-Request set at the same time as
```

Comment:
A reasonable maximum to the number of feature gates that can be activated in an epoch would be helpful (64?) as this would permit the size of these accounts, and the size of the vote account PDAs to be static

Reply:
Agreed I think this is a good idea. 64 so we can compress to `u64` somewhere?

Reply:
As per discussions on discord, we also need to include some anti-spam mechanisms in the design. Otherwise it's really easy for me to create 64 feature accounts and starve legitimate features. Even worse once revoke is supported, since I could create the accounts right before the epoch boundary, and Revoke them to reclaim the lamports a slot or two later.

I had already been pondering whether Revoke should actually burn the lamports instead of returning them. Thoughts?